From f447104228f29144f1674f9beaf16b7b175c02ca Mon Sep 17 00:00:00 2001
From: Sam Mortimer <sam@mortimer.me.uk>
Date: Sat, 26 May 2018 18:02:55 -0700
Subject: [PATCH 6/9] Add led capability
 LIGHTS_ADJUSTABLE_BATTERY_LED_BRIGHTNESS

*) Many existing liblights support notification brightness
   control via the alpha channel but do not support similar
   for the battery led.

*) https://github.com/LineageOS/android_lineage-sdk/commit/8f7a4559ab4a81855f399eae32ebe4a3f531e8b2
   introduced a dependency on light capability
   LIGHTS_ADJUSTABLE_BATTERY_LED_BRIGHTNESS such that, when
   set, it was assumed that the battery led brightness is
   also liblights adjustable.  It turned out that this
   was not the case.

*) Create a new capability to allow devices to distinguish
   whether their liblights supports adjustable brightness
   control for notification and battery leds independently
   of one another.

*) Document how brightness support works in LightsCapabilities.

*) Copy LightsCapabilities comments to the default lights config
   to raise visibility.

Change-Id: Id95905f07128c78c6acacf3b60d71bb37f70bcfb
---
 lineage/res/res/values/config.xml             | 27 +++++++++++++++++++
 .../notification/LightsCapabilities.java      | 15 ++++++++++-
 .../notification/LineageBatteryLights.java    |  2 +-
 3 files changed, 42 insertions(+), 2 deletions(-)

diff --git a/lineage/res/res/values/config.xml b/lineage/res/res/values/config.xml
index c525a65..9e961f5 100644
--- a/lineage/res/res/values/config.xml
+++ b/lineage/res/res/values/config.xml
@@ -60,14 +60,41 @@
          This integer should equal the sum of the corresponding value for each
          of the following capabilities present:
 
+         // Device has a color adjustable battery light.
          LIGHTS_RGB_NOTIFICATION_LED = 1
+
+         // Device has a color adjustable notification light.
          LIGHTS_RGB_BATTERY_LED = 2
+
          LIGHTS_MULTIPLE_NOTIFICATION_LED = 4 (deprecated)
+
+         // The notification light has adjustable pulsing capability.
          LIGHTS_PULSATING_LED = 8
+
+         // Device has a multi-segment battery light that is able to
+         // use the light brightness value to determine how many
+         // segments to show (in order to represent battery level).
          LIGHTS_SEGMENTED_BATTERY_LED = 16
+
+         // The notification light supports HAL adjustable brightness
+         // via the alpha channel.
+         // Note: if a device notification light supports LIGHTS_RGB_NOTIFICATION_LED
+         // then HAL support is not necessary for brightness control.  In this case,
+         // brightness support will be provided by lineage-sdk through the scaling of
+         // RGB color values.
          LIGHTS_ADJUSTABLE_NOTIFICATION_LED_BRIGHTNESS = 32
+
+         // Device has a battery light.
          LIGHTS_BATTERY_LED = 64
 
+         // The battery light supports HAL adjustable brightness via
+         // the alpha channel.
+         // Note: if a device battery light supports LIGHTS_RGB_BATTERY_LED then HAL
+         // support is not necessary for brightness control.  In this case,
+         // brightness support will be provided by lineage-sdk through the scaling of
+         // RGB color values.
+         LIGHTS_ADJUSTABLE_BATTERY_LED_BRIGHTNESS = 128
+
          For example, a device with notification and battery lights that supports
          pulsating and RGB control would set this config to 75. -->
     <integer name="config_deviceLightCapabilities">8</integer>
diff --git a/sdk/src/java/org/lineageos/internal/notification/LightsCapabilities.java b/sdk/src/java/org/lineageos/internal/notification/LightsCapabilities.java
index a1ee26a..81815ae 100644
--- a/sdk/src/java/org/lineageos/internal/notification/LightsCapabilities.java
+++ b/sdk/src/java/org/lineageos/internal/notification/LightsCapabilities.java
@@ -36,12 +36,25 @@ public final class LightsCapabilities {
     // segments to show (in order to represent battery level).
     public static final int LIGHTS_SEGMENTED_BATTERY_LED = 16;
 
-    // The notification light supports adjustable brightness.
+    // The notification light supports HAL adjustable brightness
+    // via the alpha channel.
+    // Note: if a device notification light supports LIGHTS_RGB_NOTIFICATION_LED
+    // then HAL support is not necessary for brightness control.  In this case,
+    // brightness support will be provided by lineage-sdk through the scaling of
+    // RGB color values.
     public static final int LIGHTS_ADJUSTABLE_NOTIFICATION_LED_BRIGHTNESS = 32;
 
     // Device has a battery light.
     public static final int LIGHTS_BATTERY_LED = 64;
 
+    // The battery light supports HAL adjustable brightness via
+    // the alpha channel.
+    // Note: if a device battery light supports LIGHTS_RGB_BATTERY_LED then HAL
+    // support is not necessary for brightness control.  In this case,
+    // brightness support will be provided by lineage-sdk through the scaling of
+    // RGB color values.
+    public static final int LIGHTS_ADJUSTABLE_BATTERY_LED_BRIGHTNESS = 128;
+
     public static boolean supports(Context context, final int capability) {
         final int capabilities = context.getResources().getInteger(
                 org.lineageos.platform.internal.R.integer.config_deviceLightCapabilities);
diff --git a/sdk/src/java/org/lineageos/internal/notification/LineageBatteryLights.java b/sdk/src/java/org/lineageos/internal/notification/LineageBatteryLights.java
index a318ce1..121ac8e 100644
--- a/sdk/src/java/org/lineageos/internal/notification/LineageBatteryLights.java
+++ b/sdk/src/java/org/lineageos/internal/notification/LineageBatteryLights.java
@@ -81,7 +81,7 @@ public final class LineageBatteryLights {
                 mContext, LightsCapabilities.LIGHTS_RGB_BATTERY_LED);
 
         mHALAdjustableBrightness = LightsCapabilities.supports(
-                mContext, LightsCapabilities.LIGHTS_ADJUSTABLE_NOTIFICATION_LED_BRIGHTNESS);
+                mContext, LightsCapabilities.LIGHTS_ADJUSTABLE_BATTERY_LED_BRIGHTNESS);
 
         // We support brightness adjustment if either the HAL supports it
         // or the light is RGB adjustable.
-- 
2.17.0

